### ISSUE TITLE
[BUG] Wrong sExtra in auto-generated contract documentation

### ISSUE THREAD

**Issue by gromakovsky** [tqtezos/baseDAO#115]
# Description

In all auto-generated documentations we have `sExtra :Natural`.
Example: https://github.com/tqtezos/baseDAO/blob/autodoc/master/TreasuryDAO.md#storage-Storage

However, different DAO contracts have different storages, and TreasryDAO specifically has something different from `Natural`.

# To Reproduce

Open auto documentation for Treasury DAO.
Observe `sExtra :Natural`.

# Expected behavior

`sExtra` has the type it actually has in code.

# Environment

Doesn't matter.

**Comment by Martoon-00** [comment]
I think this should be fixed in `dStorage`.

I propose to add some special type `newtype StorageWrapper a = StorageWrapper a` that has a special `TypeHasDoc` instance. This instance should delegate all `TypeHasDoc`'s methods implementations to implementations for `TypeHasDoc a`, except for `typeDoc{Michelson,Haskell}Rep` - for them we want to print representation of the type itself with exactly the same type arguments that are supplied to our `TypeHasDoc` instance.

Then we should use this `StorageWrapper` in `dStorage`.

**Comment by rinn7e** [comment]
> I think this should be fixed in `dStorage`.
> 
> I propose to add some special type `newtype StorageWrapper a = StorageWrapper a` that has a special `TypeHasDoc` instance. This instance should delegate all `TypeHasDoc`'s methods implementations to implementations for `TypeHasDoc a`, except for `typeDoc{Michelson,Haskell}Rep` - for them we want to print representation of the type itself with exactly the same type arguments that are supplied to our `TypeHasDoc` instance.
> 
> Then we should use this `StorageWrapper` in `dStorage`.

I don't totally understand on what to do exactly, currently I have something like this:

```haskell
dStorage :: forall store. TypeHasDoc store => DStorageType
dStorage = DStorageType $ DType (Proxy @(StorageWrapper store))

newtype StorageWrapper a = StorageWrapper a
  deriving stock (Show, Eq, Ord, Generic)
  deriving newtype (IsoValue)

instance (TypeHasDoc a) => TypeHasDoc (StorageWrapper a) where
  typeDocMdDescription = [md|TODO|]
  typeDocMdReference = poly1TypeDocMdReference
  typeDocDependencies p =
    genericTypeDocDependencies p <>
    [ dTypeDep @a
    , dTypeDep @MText, dTypeDep @Integer  -- for examples below
    ]
  typeDocHaskellRep = concreteTypeDocHaskellRep @(StorageWrapper (MText, Integer))
  typeDocMichelsonRep = concreteTypeDocMichelsonRep @(StorageWrapper (MText, Integer))
```

This is obviously not working, so I hope to get your advice on what changes need to be made when you get back.

**Comment by Martoon-00** [comment]
I think we should just have

```
  typeDocMdDescription = typeDocMdDescription @a
  typeDocMdReference = typeDocMdReference @a
  typeDocDependencies p = typeDocDependencies (Proxy @a)
  typeDocHaskellRep = concreteTypeDocHaskellRep @a
  typeDocMichelsonRep = concreteTypeDocMichelsonRep @a
```

+ make sure that in the rendered doc Michelson/Haskell representations are given not as "examples" (because `concreteTypeDoc*Rep attach "example" word).

**Comment by pasqu4le** [comment]
Note: AFAIU this issue only affects the Lorentz version.

It would be nice to fix this, but this is low priority, so let's do this only after everything else.

**Comment by pasqu4le** [comment]
Since my last comment the Lorentz version of the contract has been removed, so I'm closing this issue as it is no longer relevant.

**Comment by gromakovsky** [comment]
Note: I don't remember whether this bug was caused by something in Lorentz itself or by incorrect usage of Lorentz here. If it's the former, maybe another issue should be created in Lorentz.

**Comment by rinn7e** [comment]
> Note: I don't remember whether this bug was caused by something in Lorentz itself or by incorrect usage of Lorentz here. If it's the former, maybe another issue should be created in Lorentz.

As what have been discuss previously, I believe this has something to do with `dStorage` in lorentz and not how we use it in baseDao. 

But since lorentz-baseDao uses a very specific way of defining storage (polymorphic type storage) which causes this bug, I'm not sure how relevant it is for general lorentz usage.

**Comment by pasqu4le** [comment]
So I checked this again now and went back to the commit on `master` before the PR that deprecated the lorentz version (#180).

The `TypeHasDoc` instance for `Storage` at that point was:
```haskell
instance Each [IsoValue, TypeHasDoc] [ce, pm] => TypeHasDoc (Storage ce pm) where
   typeDocMdDescription =
     "Storage type for baseDAO contract"
   typeDocMdReference = poly2TypeDocMdReference
   typeDocHaskellRep = concreteTypeDocHaskellRep @(Storage Natural MText)
   typeDocMichelsonRep = concreteTypeDocMichelsonRep @(Storage Natural MText)
```
the first odd thing I noticed is the `@(Storage Natural MText)` in the last two lines, but changing it to `@(Storage ce pm)` resulted in overlapping instances issues, which I suppose is the problem at hand.

However this is quickly solved by switching to use `homomorphicTypeDocHaskellRep` and `homomorphicTypeDocMichelsonRep`, which seems to be applicable here.

I pushed the changes as proof in [a branch](https://github.com/tqtezos/baseDAO/tree/pasqu4le/%23115-tmp) and from the [pipeline results](https://buildkite.com/serokell/basedao/builds/2255#4f36c45b-851b-4509-92a9-df8b293dca1e) it seems to be working fine. From the `TreasuryDAO.md` example we get:
```
  * ***sExtra*** :[`TreasuryDaoContractExtra`](#types-TreasuryDaoContractExtra)
```
instead of
```
  * ***sExtra*** :[`Natural`](#types-Natural)
```


One thing to note is also the fact that the `Michelson.Typed.Haskell.Doc` has now changes quite a bit, but I didn't try a newer version because it would take too much to bump `morley` from that old commit of baseDAO.
It looks like even without using `homomorphicType*` it should be possible to get around the issue, although maybe not as comfortably. 
All things considered for now I don't think an issue in `morley` is necessary.
